source/common/router/config_impl.cc

```diff
@@ -1799,6 +1796,9 @@ VirtualHostImpl::VirtualHostImpl(
   }
 
   if (!virtual_host.virtual_clusters().empty()) {
+    vcluster_scope_ = Stats::Utility::scopeFromStatNames(
```

Comment:
I think it may be better to move into constructor initializer list, and use the ternary operator? In this way the vcluster_scope_ can be made const (which it should have been in the first place).

Reply:
That brings a question... Does it even need to be a class member?

Reply:
I see your point, but looking at this, I think I'd vote against it.

Instead, what we should actually do (IMO) is drop stat_name_storage_ as a member variable altogether, and create that stat-name only in this if-block.

Reply:
the scope needs to be a class member as that's what dictates the lifetime of the stats.

When the shared ref to the scope is dropped all the stats created in it will be deleted.

Reply:
Having said that, I think having all vcluster-related initializations happen in one place is more readable/discoverable. But if you feel strongly about it being const, I can move it back to the c'tor initializer list.

Reply:
You can do that, but like I said above -- you'll need to use a static helper method.

That will give you readability and also population in the initializer.

Reply:
Oh, so `stat_name_storage_` does not need to be created either? And does not need to live for the lifetime of the vhost?

Reply:
This isn't obvious to me, since `stat_name_storage_` is used in vhost's `statName()` API...

Reply:
I missed that we exposed that to end-users, and that it's possible to have a non-empty vhost-name but empty clusters.

QQ: how often does that actually happen?

Reply:
It happens most of the time. vhosts typically (always?) have names. virtual clusters are an advanced feature. I don't know how widely it's used.

Reply:
OK I'm back to thinking what you have now looks good :)

Reply:
Thanks, resolving this thread.